### PR TITLE
Add support for writing timestamp_ntz type in Delta

### LINKED ISSUE #16927: Add support for timestamp without time zone (TIMESTAMP_NTZ) type in Delta Lake connector
https://github.com/delta-io/delta/blob/master/PROTOCOL.md#timestamp-without-timezone-timestampntz

- [x] Read https://github.com/trinodb/trino/pull/17502
- [ ] Write #18913

Note: Databricks 12.2 LTS and OSS Delta Lake version 2.3.0 don't include this type. We need to wait for the next release.

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Add support for writing timestamp_ntz type in Delta

Fixes #16927

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Add support for writing `timestamp` type. ({issue}`16927`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 1]
```diff
@@ -879,10 +883,12 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
                 ImmutableList.Builder<String> columnNames = ImmutableList.builderWithExpectedSize(tableMetadata.getColumns().size());
                 ImmutableMap.Builder<String, Object> columnTypes = ImmutableMap.builderWithExpectedSize(tableMetadata.getColumns().size());
                 ImmutableMap.Builder<String, Map<String, Object>> columnsMetadata = ImmutableMap.builderWithExpectedSize(tableMetadata.getColumns().size());
+                boolean containsTimestampType = false;
                 for (ColumnMetadata column : tableMetadata.getColumns()) {
                     columnNames.add(column.getName());
                     columnTypes.put(column.getName(), serializeColumnType(columnMappingMode, fieldId, column.getType()));
                     columnsMetadata.put(column.getName(), generateColumnMetadata(columnMappingMode, fieldId));
+                    containsTimestampType |= containsTimestampType(column.getType());
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:891]: no need to call `containsTimestampType` when `containsTimestampType` is already true

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 2]
```diff
@@ -1156,6 +1164,21 @@ private static void deleteRecursivelyIfExists(TrinoFileSystem fileSystem, Locati
...
+        if (type instanceof ArrayType arrayType) {
+            return containsTimestampType(arrayType.getElementType());
+        }
+        if (type instanceof MapType mapType) {
+            return containsTimestampType(mapType.getKeyType()) || containsTimestampType(mapType.getValueType());
+        }
+        if (type instanceof RowType rowType) {
+            return rowType.getFields().stream().anyMatch(field -> containsTimestampType(field.getType()));
+        }
+        checkState(type.getTypeParameters().isEmpty(), "Type parameter should be empty: %s", type);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1178]: `checkArgument("Unexpected type parameters for type %s", type);`

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 3]
```diff
@@ -2258,7 +2283,17 @@ private ProtocolEntry protocolEntryForNewTable(Map<String, Object> properties)
...
+        if (containsTimestampType) {
+            readerVersion = max(readerVersion, TIMESTAMP_NTZ_SUPPORTED_READER_VERSION);
+            writerVersion = max(writerVersion, TIMESTAMP_NTZ_SUPPORTED_WRITER_VERSION);
+            readerFeatures.add("timestampNtz");
+            writerFeatures.add("timestampNtz");
+        }
+        return new ProtocolEntry(
+                readerVersion,
+                writerVersion,
+                readerFeatures.isEmpty() ? Optional.empty() : Optional.of(readerFeatures),
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2295]: Why discern between empty Set of features and no set of features?

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 3 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2297]: `ProtocolEntry` has the following logic, so `Optional.of(Set.of())` throws an exception for lower reader & writer versions. We could change the logic alternatively. 
```java
        if (minReaderVersion < MIN_VERSION_SUPPORTS_READER_FEATURES && readerFeatures.isPresent()) {
            throw new IllegalArgumentException("readerFeatures must not exist when minReaderVersion is less than " + MIN_VERSION_SUPPORTS_READER_FEATURES);
        }
        if (minWriterVersion < MIN_VERSION_SUPPORTS_WRITER_FEATURES && writerFeatures.isPresent()) {
            throw new IllegalArgumentException("writerFeatures must not exist when minWriterVersion is less than " + MIN_VERSION_SUPPORTS_WRITER_FEATURES);
        }
```

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

**Pull request merged**